torch_geometric/data/lightning_datamodule.py

```diff
@@ -481,11 +487,36 @@ def __init__(
 
         if loader in ['neighbor', 'link_neighbor']:
             input_type = get_edge_label_index(data, input_train_edges)[0]
-            if input_type is not None:
-                num_src_nodes = data[input_type[0]].num_nodes
-                num_dst_nodes = data[input_type[-1]].num_nodes
+
+            if isinstance(data, (Data, HeteroData)):
+                if input_type is not None:
+                    num_src_nodes = data[input_type[0]].num_nodes
+                    num_dst_nodes = data[input_type[-1]].num_nodes
+                else:
+                    num_src_nodes = num_dst_nodes = data.num_nodes
             else:
-                num_src_nodes = num_dst_nodes = data.num_nodes
+                # TODO improve num_nodes logic, currently we can only do this
+                # by explicitly querying a feature tensor
```

Comment:
why? I thought that we still expose num_nodes in table data.

Reply:
this is the case of a `FeatureStore` generic object, which does not currently support `num_nodes`

Reply:
Can we add this to `FeatureStore`? I feel we are replicating this logic all over PyG.

Reply:
I think we can, but I would prefer to do that in a separate PR. The best way to incorporate this would be to have a RemoteBackend class that performs operations on a Tuple[FeatureStore, GraphStore], one of which is `RemoteBackend.num_nodes(FeatureStore, GraphStore)`. 

I can implement this in a separate PR and get rid of all the cases in PyG where we introduce this logic, but it may be a good idea to implement this first for correctness. Wdyt?

Reply:
Does it need to be so complex? I think it makes sense to just support `GraphStore.num_src_nodes(edge_type)` and `GraphStore.num_dst_nodes(edge_type)`. I am not 100% convinced we should use the `FeatureStore` for this at all.

Reply:
We also have `EdgeAttr.size` we can leverage for this, right?

Reply:
Yes, looks like `EdgeAttr` is easier.

Reply:
Yes, this interface works. It may still be worth considering inferring based on features as well for link prediction tasks where the input edge type is not a part of the graph. For this case, we are passing-through kwargs as before.